pkg/gencosmosdb/cosmosdb/authorizer.go

```diff
@@ -39,13 +42,96 @@ func NewMasterKeyAuthorizer(masterKey string) (Authorizer, error) {
 }
 
 type tokenAuthorizer struct {
-	token string
+	token       string
+	expiration  time.Time
+	cond        *sync.Cond
+	acquiring   bool
+	lastAttempt time.Time
+	getToken    func() (token string, newExpiration time.Time, err error)
+}
+
+func (a *tokenAuthorizer) Authorize(req *http.Request, resourceType, resourceLink string) error {
+	token, err := a.aquireToken()
+	if err != nil {
+		return err
+	}
+	authorizationHeader := fmt.Sprintf("type=aad&ver=1.0&sig=%s", token)
+	currentTime := time.Now().UTC().Format("Mon, 02 Jan 2006 15:04:05 GMT")
+	req.Header.Set("Authorization", authorizationHeader)
+	req.Header.Set("x-ms-date", currentTime)
+
+	return nil
 }
 
-func (a *tokenAuthorizer) Authorize(req *http.Request, resourceType, resourceLink string) {
-	req.Header.Set("Authorization", url.QueryEscape(a.token))
+func NewTokenAuthorizer(token string, expiration time.Time, getToken func() (token string, newExpiration time.Time, err error)) Authorizer {
+	return &tokenAuthorizer{token: token, expiration: expiration, getToken: getToken, cond: sync.NewCond(&sync.Mutex{})}
 }
 
-func NewTokenAuthorizer(token string) Authorizer {
-	return &tokenAuthorizer{token: token}
+// Get returns the underlying resource.
+// If the resource is fresh, no refresh is performed.
+func (a *tokenAuthorizer) aquireToken() (string, error) {
+	// If the resource is expiring within this time window, update it eagerly.
+	// This allows other goroutines to keep running by using the not-yet-expired
+	// resource value while one goroutine updates the resource.
+	const window = 5 * time.Minute   // To update the token if 5 mins left to expire
+	const backoff = 30 * time.Second // Minimum wait time between attempts
+	now, acquire, expired := time.Now(), false, false
+
+	// acquire exclusive lock
+	a.cond.L.Lock()
+	token := a.token
+
+	for {
```

Comment:
I think the majority of this for loop can be re-organised and simplified.

Reply:
Since this is a little complex structure, those extra lines are needed to have proper comments for someone to understand. For example if you see the `if` statement in the `for` loop, the middle `else if` and the last `else` both have the lines -

> 			token = a.token
>			break

However if you see the comments, both the lines are there to use the existing token but for different reasons.